### PR TITLE
WIP: change pytest setup

### PR DESCRIPTION
**Pull request by kohr-h**
This is a "discussion PR" for issue #1147 regarding conflicts of pytest plugins.

**Current status:** Requires `pytest odl` to run tests. Just `pytest` fails.

What I think we should do irrespective of the overall issue is to move the `test_examples.py` and `test_doc.py` out of `odl/test` to the `examples/` and `doc/`directories, respectively, and supply them with a custom `conftest.py`. That way, testing only the examples or only the docs becomes easier, namely one only needs `pytest examples` and `pytest doc`, respectively, without the extra `--examples` or `--doctest-doc` command line options.

### CODE REVIEW

File: doc/conftest.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+from odl.trafos.backends import PYFFTW_AVAILABLE, PYWT_AVAILABLE
+from odl.tomo.backends import (
+    ASTRA_AVAILABLE, ASTRA_CUDA_AVAILABLE, SKIMAGE_AVAILABLE)
+
+# We don't bother excluding stuff depending on available backends. We just
+# require them all (except stir)
+if not all([PYFFTW_AVAILABLE, PYWT_AVAILABLE,
+            ASTRA_AVAILABLE, ASTRA_CUDA_AVAILABLE, SKIMAGE_AVAILABLE]):
+    print('to be able to test the docs, you need to install all backends, '
```
Comment (adler-j) [inline doc/conftest.py]: Why not just raise an exception?

File: doc/conftest.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+import os
+import sys
+
+from odl.trafos.backends import PYFFTW_AVAILABLE, PYWT_AVAILABLE
+from odl.tomo.backends import (
+    ASTRA_AVAILABLE, ASTRA_CUDA_AVAILABLE, SKIMAGE_AVAILABLE)
+
+# We don't bother excluding stuff depending on available backends. We just
+# require them all (except stir)
```
Comment (adler-j) [inline doc/conftest.py]: We should just move stir to contrib asap :/

File: doc/conftest.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    print('to be able to test the docs, you need to install all backends, '
+          'currently: '
+          '`astra`, `pyfftw`, `pywt`, `skimage`; '
+          'please consult the installation documentation for details',
+          file=sys.stderr)
+    sys.exit(-1)
+
+here = os.path.normpath(os.path.dirname(__file__))
+collect_ignore = [os.path.join(here, 'sphinxext')]
+collect_ignore = [os.path.normcase(ignored) for ignored in collect_ignore]
```
Comment (adler-j) [inline doc/conftest.py]: Dislike overwriting names like this.

File: doc/test_doc.py [hunk 4]
```diff
@@ -39,9 +39,8 @@
 doctest_extraglobs = {'odl': odl, 'np': numpy, 'scipy': scipy,
```
Comment (adler-j) [inline doc/test_doc.py]: Is this file still needed if we move to `pytest doc`?

File: examples/conftest.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
+# Copyright 2014-2017 The ODL contributors
+#
+# This file is part of ODL.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+
+"""Test configuration file."""
```
Comment (adler-j) [inline examples/conftest.py]: This needs to be improved. Also add a mention of this file to the readme

File: doc/conftest.py (hunk 3 above)
Comment (kohr-h) [reply doc/conftest.py]: What's wrong with it? When it's 100 lines apart, okay, but two consecutive lines? Why use a new name?

Comment (adler-j) [reply doc/conftest.py]:
```
collect_ignore = [os.path.normcase(os.path.join(here, 'sphinxext'))]
```

whats wrong with that?

Comment (kohr-h) [reply doc/conftest.py]: Nothing, but how about other files with more entries in this list? Use `os.path.normcase(os.path.join())` for each entry? Seems like a lot of noise to me.
